File: sdk-java/src/test/java/io/superdurable/dex/integ/NoStartStateTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Tag("dex-dev")
public final class NoStartStateTest {
    private static final NoStartStateWorkflow NO_START_WORKFLOW = new NoStartStateWorkflow();
    private static final RpcNoStateWorkflow NO_STEP_WORKFLOW = new RpcNoStateWorkflow();
    private static final NoStartStateDeadEndWorkflow DEAD_END_WORKFLOW =
            new NoStartStateDeadEndWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testNoStartStateWorkflow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                NO_START_WORKFLOW)) {
            final String flowId = "no-start-" + UUID.randomUUID();
            environment.client().startFlow(NO_START_WORKFLOW, flowId, null);
            final NoStartStateWorkflow stub = environment.client().newRpcStub(
                    NoStartStateWorkflow.class,
                    flowId);
            assertEquals(
                    NoStartStateWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::invoke, "rpc-input"));
            assertEquals(1, environment.client().waitForFlow(flowId, Duration.ofSeconds(30)).getSingleOutput(Integer.class));
        }
    }

    @Test
    void testNoStateWorkflow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                NO_STEP_WORKFLOW)) {
            final String flowId = "no-state-" + UUID.randomUUID();
            environment.client().startFlow(NO_STEP_WORKFLOW, flowId, null);
            final RpcNoStateWorkflow stub = environment.client().newRpcStub(
                    RpcNoStateWorkflow.class,
                    flowId);
            assertEquals(
                    RpcNoStateWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::invoke, "rpc-input"));
            environment.client().stopFlow(flowId);
        }
    }

    @Test
    void testDeadEndWorkflow() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                DEAD_END_WORKFLOW)) {
            final String flowId = "dead-end-" + UUID.randomUUID();
            environment.client().startFlow(DEAD_END_WORKFLOW, flowId, null);
            final NoStartStateDeadEndWorkflow stub = environment.client().newRpcStub(
                    NoStartStateDeadEndWorkflow.class,
                    flowId);
            assertEquals(
                    NoStartStateDeadEndWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::invoke, "rpc-input"));
            assertTrue(environment.client()
                    .waitForFlow(flowId, Duration.ofSeconds(30))
                    .getCompletions()
                    .isEmpty());
        }
    }

    void compileNoStartStep(final Client client) {
        client.startFlow(NO_START_WORKFLOW, "no-start", null);
        final NoStartStateWorkflow stub = client.newRpcStub(
                NoStartStateWorkflow.class,
                "no-start");
        final Long output = client.invokeRPC(stub::invoke, "input");
        consume(output);
    }

    void compileNoStep(final Client client) {
        client.startFlow(NO_STEP_WORKFLOW, "no-step", null);
        final RpcNoStateWorkflow stub = client.newRpcStub(
                RpcNoStateWorkflow.class,
                "no-step");
        final Integer output = client.invokeRPC(stub::increaseCounter);
        client.stopFlow("no-step");
        consume(output);
    }

    void compileDeadEnd(final Client client) {
        client.startFlow(DEAD_END_WORKFLOW, "dead-end", null);
        final NoStartStateDeadEndWorkflow stub = client.newRpcStub(
                NoStartStateDeadEndWorkflow.class,
                "dead-end");
        final Integer size = client.invokeRPC(stub::publishInternal);
        consume(size);
    }

    private static void consume(final Object value) {
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/no_start_state_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::time::Duration;

use dex_sdk::{Client, Registry, SdkResult, StopFlowOptions};

use crate::no_start_state_dead_end_workflow::NoStartStateDeadEndWorkflow;
use crate::no_start_state_workflow::NoStartStateWorkflow;
use crate::rpc_no_state_workflow::RpcNoStateWorkflow;
use crate::support::{DexDevTestEnvironment, flow_id};

#[test]
#[ignore = "requires dexcli dev"]
fn test_no_start_state_workflow() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(NoStartStateWorkflow::new()));
    let workflow = NoStartStateWorkflow::new();
    let flow_id = flow_id("no-start");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start Flow without a start Step");
    assert_eq!(
        NoStartStateWorkflow::RPC_OUTPUT,
        environment
            .client
            .invoke_rpc(
                &flow_id,
                NoStartStateWorkflow::INVOKE,
                "rpc-input".to_string(),
            )
            .expect("invoke no-start RPC")
    );
    assert_eq!(
        1,
        environment
            .client
            .wait_for_flow_with_timeout::<i32>(&flow_id, Duration::from_secs(30))
            .expect("complete no-start Flow")
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_no_state_workflow() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(RpcNoStateWorkflow::new()));
    let workflow = RpcNoStateWorkflow::new();
    let flow_id = flow_id("no-state");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start no-State Flow");
    assert_eq!(
        RpcNoStateWorkflow::RPC_OUTPUT,
        environment
            .client
            .invoke_rpc(
                &flow_id,
                RpcNoStateWorkflow::INVOKE,
                "rpc-input".to_string(),
            )
            .expect("invoke no-State RPC")
    );
    environment
        .client
        .stop_flow(&flow_id, StopFlowOptions::cancel())
        .expect("stop no-State Flow");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_dead_end_workflow() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(NoStartStateDeadEndWorkflow::new()));
    let workflow = NoStartStateDeadEndWorkflow::new();
    let flow_id = flow_id("dead-end");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start dead-end Flow");
    assert_eq!(
        100,
        environment
            .client
            .invoke_rpc(
                &flow_id,
                NoStartStateDeadEndWorkflow::INVOKE,
                "rpc-input".to_string(),
            )
            .expect("resume dead-end Flow")
    );
    environment
        .client
        .wait_for_flow_with_timeout::<()>(&flow_id, Duration::from_secs(30))
        .expect("complete resumed dead-end Flow");
}

#[allow(dead_code)]
fn compile_no_start_step(client: &Client) -> SdkResult<()> {
    client.start_flow(&NoStartStateWorkflow::new(), "no-start", ())?;
    let _: i64 = client.invoke_rpc(
        "no-start",
        NoStartStateWorkflow::INVOKE,
        "input".to_string(),
    )?;
    Ok(())
}

#[allow(dead_code)]
fn compile_no_step(client: &Client) -> SdkResult<()> {
    client.start_flow(&RpcNoStateWorkflow::new(), "no-step", ())?;
    let _: i32 =
        client.invoke_rpc_without_input("no-step", RpcNoStateWorkflow::INCREASE_COUNTER)?;
    client.stop_flow("no-step", StopFlowOptions::cancel())
}

#[allow(dead_code)]
fn compile_dead_end(client: &Client) -> SdkResult<()> {
    client.start_flow(&NoStartStateDeadEndWorkflow::new(), "dead-end", ())?;
    let _: usize = client
        .invoke_rpc_without_input("dead-end", NoStartStateDeadEndWorkflow::PUBLISH_INTERNAL)?;
    Ok(())
}
